torch/csrc/distributed/c10d/symm_mem/NVSHMEMSymmetricMemory.cu

```diff
@@ -261,7 +262,7 @@ class NVSHMEMSymmetricMemoryAllocator : public SymmetricMemoryAllocator {
     auto allocation =
         std::make_shared<NVSHMEMAllocation>(ptr, size, device_idx);
     // TODO: thread safety
-    allocations_.emplace(ptr, allocation);
+    allocations_.try_emplace(ptr, std::move(allocation));
```

Comment:
Efficient and remove an atomic lock. Still think this could be a unique ptr though.